stripe/src/test/java/com/stripe/android/StripeKtxTest.kt

```diff
@@ -0,0 +1,552 @@
+            ).thenReturn(null)
+
+            assertFailsWith<InvalidRequestException> {
+                stripe.confirmPaymentIntent(mock())
+            }
+        }
+
+    private inline fun <reified APIObject : Any, reified CreateAPIParam : Any, reified RepositoryParam : Any>
+    `Given repository returns non-empty value when calling createAPI then returns correct result`(
+        crossinline repositoryInvocationBlock: suspend (RepositoryParam, ApiRequest.Options) -> APIObject?,
+        crossinline createAPIInvocationBlock: suspend (CreateAPIParam, String?, String?) -> APIObject
+    ) = runBlocking {
+        val expectedApiObj = mock<APIObject>()
+
+        whenever(
+            repositoryInvocationBlock(any(), any())
+        ).thenReturn(expectedApiObj)
+
+        val actualObj = createAPIInvocationBlock(
+            mock(),
+            TEST_IDEMPOTENCY_KEY,
+            TEST_STRIPE_ACCOUNT_ID
+        )
+
+        assertSame(expectedApiObj, actualObj)
+    }
+
+    private inline fun <APIObject : Any, reified CreateAPIParam : Any, reified RepositoryParam : Any>
+    `Given repository throws exception when calling createAPI then throws same exception`(
+        crossinline repositoryInvocationBlock: suspend (RepositoryParam, ApiRequest.Options) -> APIObject?,
+        crossinline createAPIInvocationBlock: suspend (CreateAPIParam, String?, String?) -> APIObject
+    ): Unit = runBlocking {
+        whenever(
+            repositoryInvocationBlock(any(), any())
+        ).thenThrow(mock<AuthenticationException>())
+
+        assertFailsWith<AuthenticationException> {
+            createAPIInvocationBlock(
+                mock(),
+                TEST_IDEMPOTENCY_KEY,
+                TEST_STRIPE_ACCOUNT_ID
+            )
+        }
+    }
+
+    private inline fun <APIObject : Any, reified CreateAPIParam : Any, reified RepositoryParam : Any>
+    `Given repository returns empty value when calling createAPI then thorws APIException`(
+        crossinline repositoryInvocationBlock: suspend (RepositoryParam, ApiRequest.Options) -> APIObject?,
+        crossinline createAPIInvocationBlock: suspend (CreateAPIParam, String?, String?) -> APIObject
+    ): Unit = runBlocking {
+        whenever(
+            repositoryInvocationBlock(any(), any())
+        ).thenReturn(null)
+
+        assertFailsWith<InvalidRequestException> {
+            createAPIInvocationBlock(
+                mock(),
+                TEST_IDEMPOTENCY_KEY,
+                TEST_STRIPE_ACCOUNT_ID
+            )
+        }
+    }
+
+    private inline fun <reified APIObject : Any, reified RepositoryParam : Any>
+    `Given repository returns non-empty value when calling createAPI with String param then returns correct result`(
+        crossinline repositoryInvocationBlock: suspend (RepositoryParam, ApiRequest.Options) -> APIObject?,
+        crossinline createAPIInvocationBlock: suspend (String, String?, String?) -> APIObject
```

Comment:
It feels like the distinction between "create" and "retrieve" methods isn't super important. Can we simplify the methods?

Reply:
`createAPIInvocationBlock` and `retrieveAPIInvocationBlock` takes different parameters, I think the only way to merge them into one is for the template function to take the entire block to actually call the function, like
`
template(
mockApiRepository::retrieveSetupIntent,
{stripe.retrieveCall(param1, param2)}
`
`
template(
mockApiRepository::retrieveSetupIntent,
{stripe.createCall(param1, param2, param3)}
`
but that way the actual test will have a lot of duplicate params, which doesn't necessarily make it more readable